sage.py

```diff
@@ -104,105 +89,72 @@ def load_IANA_mapping():
-#ips = dict()
-#hosts = dict()
-#h_trig = []
 
-def parse(unparsed_data, alert_labels=[], slim=False):
-    
-    FILTER = False
+# Step 1.1: Parse the input alerts
+def _parse(unparsed_data, filter_alerts=False):
     badIP = '169.254.169.254'
     __cats = set()
     __ips = set()
     __hosts = set()
-    __sev = set()
-    data = []
-    connections = dict()
+    parsed_data = []
 
     prev = -1
-    for id, d in enumerate(unparsed_data):
-        #print(d)
-        raw = ''
-        
-        try:
-             raw = json.loads(d['result']['_raw'])
-        except:
-            try:
-                raw = json.loads(d['_raw'])
-            except:
-                raw = d
-                
+    for d in unparsed_data:
+        if 'result' in d and '_raw' in d['result']:
+            raw = json.loads(d['result']['_raw'])
+        elif '_raw' in d:
+            raw = json.loads(d['_raw'])
+        else:
+            raw = d
+
         if raw['event_type'] != 'alert':
             continue
-        #app_proto = raw['app_proto']
-        host = ''
-        
-        try:
+
+        if 'host' in raw:
             host = raw['host']
-        except:
-            try:
-                host = d['host'][3:]
-            except:
-                host = 'dummy'
-                
-
-        #print(host)
-        ts = raw['timestamp']
-        dt = datetime.datetime.strptime(ts, '%Y-%m-%dT%H:%M:%S.%f%z')# 2018-11-03T23:16:09.148520+0000
-        DIFF = 0.0 if prev== -1 else round((dt - prev).total_seconds(),2)
+        elif 'host' in d:
+            host = d['host'][3:]
+        else:
+            host = 'dummy'
+
+        dt = datetime.datetime.strptime(raw['timestamp'], '%Y-%m-%dT%H:%M:%S.%f%z')  # 2018-11-03T23:16:09.148520+0000
+        diff_dt = 0.0 if prev == -1 else round((dt - prev).total_seconds(), 2)
         prev = dt
-        
-                
+
         sig = raw['alert']['signature']
         cat = raw['alert']['category']
-        
-        severity = raw['alert']['severity']
 
-        if cat == 'Attempted Information Leak' and FILTER:
+        # Filter out the alert that occurs way too often
+        if cat == 'Attempted Information Leak' and filter_alerts:
             continue
-        srcip = raw['src_ip']
-        srcport = None if 'src_port' not in raw.keys() else raw['src_port']
-        dstip = raw['dest_ip']
-        dstport = None if 'dest_port' not in raw.keys() else raw['dest_port']
 
-        # Filtering out mistaken alerts / uninteresting alerts
-        if srcip == badIP or dstip == badIP or cat == 'Not Suspicious Traffic':
+        src_ip = raw['src_ip']
+        src_port = None if 'src_port' not in raw.keys() else raw['src_port']
+        dst_ip = raw['dest_ip']
+        dst_port = None if 'dest_port' not in raw.keys() else raw['dest_port']
+
+        # Filter out mistaken alerts / uninteresting alerts
+        if src_ip == badIP or dst_ip == badIP or cat == 'Not Suspicious Traffic':  # TODO: add `filter_alerts` here?
```

Comment:
As already discussed, I removed the "slim" part. Do you want to add boolean`filter` here? This might change the resulting AGs though

Reply:
I have only written it in the TODO, the flags are the same as they were before

Reply:
I have created an issue for this part, you might check it and comment if necessary. I will resolve all other comments related to this issue